pkg/apis/core/validation/ratcheting.go

```diff
@@ -0,0 +1,124 @@
+	"k8s.io/kubernetes/pkg/apis/core"
+)
+
+var validStorageMedia = sets.NewString(
+	string(core.StorageMediumDefault),
+	string(core.StorageMediumMemory),
+	string(core.StorageMediumHugePages),
+)
+
+// RatchetingPodSpecValidations is a set of additional validations added to validate pod specs.
+// This should be passed to ValidateRatchetingCreate and ValidateRatchetingUpdate along with *core.PodSpec data to validate.
+var RatchetingPodSpecValidations = apimachineryvalidation.RatchetingValidators{
+	// flex volume source driver name
+	func(data interface{}, fldPath *field.Path) field.ErrorList {
+		allErrs := field.ErrorList{}
+		spec := data.(*core.PodSpec)
+		for i, v := range spec.Volumes {
+			if v.VolumeSource.FlexVolume != nil && len(v.VolumeSource.FlexVolume.Driver) > 0 {
+				for _, msg := range validation.IsQualifiedName(v.VolumeSource.FlexVolume.Driver) {
+					allErrs = append(allErrs, field.Invalid(fldPath.Child("volumes").Index(i).Child("flexVolume", "driver"), v.VolumeSource.FlexVolume.Driver, msg))
+				}
+			}
+		}
+		return allErrs
+	},
+
+	// emptyDir storage medium
+	func(data interface{}, fldPath *field.Path) field.ErrorList {
+		allErrs := field.ErrorList{}
+		spec := data.(*core.PodSpec)
+		for i, v := range spec.Volumes {
+			if v.VolumeSource.EmptyDir != nil {
+				if !validStorageMedia.Has(string(v.VolumeSource.EmptyDir.Medium)) {
+					allErrs = append(allErrs, field.NotSupported(fldPath.Child("volumes").Index(i).Child("emptyDir", "medium"), v.VolumeSource.EmptyDir.Medium, validStorageMedia.List()))
+				}
+			}
+		}
+		return allErrs
+	},
+
+	// duplicate envvar names
+	func(data interface{}, fldPath *field.Path) field.ErrorList {
+		allErrs := field.ErrorList{}
+		visitContainers(data.(*core.PodSpec), fldPath, func(c *core.Container, fldPath *field.Path) {
+			names := sets.NewString()
+			for j, e := range c.Env {
+				if names.Has(e.Name) {
+					allErrs = append(allErrs, field.Duplicate(fldPath.Child("env").Index(j).Child("name"), e.Name))
+				} else {
+					names.Insert(e.Name)
+				}
+			}
+		})
+		return allErrs
+	},
+
+	// duplicate pvc volumes
+	func(data interface{}, fldPath *field.Path) field.ErrorList {
+		allErrs := field.ErrorList{}
+		spec := data.(*core.PodSpec)
+		names := sets.NewString()
+		for i, v := range spec.Volumes {
+			if pvc := v.VolumeSource.PersistentVolumeClaim; pvc != nil {
+				if names.Has(pvc.ClaimName) {
+					allErrs = append(allErrs, field.Duplicate(fldPath.Child("volumes").Index(i).Child("persistentVolumeClaim").Child("claimName"), pvc.ClaimName))
+				} else {
+					names.Insert(pvc.ClaimName)
+				}
+			}
+		}
+		return allErrs
+	},
+
+	// memory units
+	func(data interface{}, fldPath *field.Path) field.ErrorList {
+		allErrs := field.ErrorList{}
+		visitContainers(data.(*core.PodSpec), fldPath, func(c *core.Container, fldPath *field.Path) {
+			for k, q := range c.Resources.Limits {
+				if k == core.ResourceMemory && q.MilliValue()%1000 != 0 {
```

Comment:
Comment on `MilliValue` notes:
> // MilliValue returns the value of ceil(q * 1000); this could overflow an int64;
// if that's a concern, call Value() first to verify the number is small enough.

Can you just do: `_, loss := q.AsScale(0); loss` ?

Reply:
done (though the loss bool was flipped)